core/src/price_estimation/priority_price_source.rs

```diff
@@ -0,0 +1,143 @@
+use super::PriceSource;
+use crate::models::TokenId;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::iter::FromIterator;
+
+/**
+ * A price source that sequentially queries its inner sources in order and returns the
+ * first price found.
+ */
+pub struct PriorityPriceSource {
+    sources: Vec<Box<dyn PriceSource + Send + Sync>>,
+}
+
+impl PriorityPriceSource {
+    #[allow(dead_code)]
+    pub fn new(sources: Vec<Box<dyn PriceSource + Send + Sync>>) -> Self {
+        Self { sources }
+    }
+}
+
+impl PriceSource for PriorityPriceSource {
+    fn get_prices<'a>(
+        &'a self,
+        tokens: &'a [TokenId],
+    ) -> BoxFuture<'a, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let mut remaining_tokens: HashSet<TokenId, RandomState> =
```

Comment:
nit: `RandomState` is the default hash state for the a `HashSet` so it is superfluous here.
```suggestion
            let mut remaining_tokens: HashSet<TokenId> =
```

Also, the type for the `HashSet` item can also be elided in this case:
```suggestion
            let mut remaining_tokens: HashSet<_> =
```